### PR TITLE
TIKA-4126

### PR DESCRIPTION
**Pull request by tballison**
Thanks for your contribution to [Apache Tika](https://tika.apache.org/)! Your help is appreciated!

Before opening the pull request, please verify that
* there is an open issue on the [Tika issue tracker](https://issues.apache.org/jira/projects/TIKA) which describes the problem or the improvement. We cannot accept pull requests without an issue because the change wouldn't be listed in the release notes.
* the issue ID (`TIKA-XXXX`)
  - is referenced in the title of the pull request
  - and placed in front of your commit messages surrounded by square brackets (`[TIKA-XXXX] Issue or pull request title`)
* commits are squashed into a single one (or few commits for larger changes)
* Tika is successfully built and unit tests pass by running `mvn clean test`
* there should be no conflicts when merging the pull request branch into the *recent* `main` branch. If there are conflicts, please try to rebase the pull request branch on top of a freshly pulled `main` branch
* if you add new module that downstream users will depend upon add it to relevant group in `tika-bom/pom.xml`.

We will be able to faster integrate your pull request if these conditions are met. If you have any questions how to fix your problem or about using Tika in general, please sign up for the [Tika mailing list](http://tika.apache.org/mail-lists.html). Thanks!

### CODE REVIEW

**Comment by patrickdalla** [comment]
"By the time someone passes DateUtils a Calendar, there's already a timezone on it" from tika-4126 ISSUE:

Yes, but if you see the Calendar code, this timezone that was read from XMP metadata does not immediately triggers any method to shift the hour of day field. Neither this setTimezone method, nor the "date.setTimeZone(UTC);" method in line 92 of DateUtils formatDate method.

What really triggers the field hour of day sync, is the first "calendar.get(Calendar.YEAR)" that is called in doFormatDate method (line 116). But when this is triggered, the timezone is set as UTC, ignoring the first setTimezone called when the XMP metadata was read.

Just add "date.get(Calendar.DAY_OF_MONTH)" right before line 92 of DateUtils formatDate and it will force sync the XMP metadata timezone read, so, in subsequent calendar.get(Calendar.DAY_OF_MONTH) done at line 117, the shift of the hour of day (and other fields) will be correctly done.

**Comment by patrickdalla** [comment]
"so we may as well just rely on Instant's toString(), truncated to seconds?"

I don't think this helps anyhow on solving the tika-4126 ISSUE.

**Comment by tballison** [comment]
> I don't think this helps anyhow on solving the tika-4126 ISSUE.

The unit test that I added shows that it does help, I think?  Also, I had to fix 3 other unit tests after making this correction.

The documentation here (https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/time/Instant.html#toString()) and https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/time/Instant.html#toString() suggests that Java is correctly converting the times to UTC in Instant's toString method?

Please let me know if the unit tests are wrong or if I'm misunderstanding Java or the original problem.  Thank you!

**Comment by patrickdalla** [comment]
Yes, it seems you misunderstood (or I misinformed) the issue.

There is a XMP metadata tag in the PDF with the values:
<rdf:Description rdf:about="" **xmp:ModifyDate="2023-09-06T13:35:38-04:00"** xmlns:xmp="http://ns.adobe.com/xap/1.0/">

Look that the timezone is -04:00. But tika does not extracts the timezone, parsing the value as "2023-09-06T13:35:38Z", as it was in UTC.

**Comment by patrickdalla** [comment]
Look at the code execution on the following video I have done:
https://github.com/apache/tika/assets/28692427/63f7cf7a-11b2-40e0-9720-6af181b8c8b3

If I uncomment the Calendar get method, the preceding setTimezone is correctly applied, and the following setTimezone to UTC and get methods work well, shifting the hour_of_day field correctly.

If it is kept commented, when the setTimezone to UTC is called, the following get method does not shift the internal fields as it considers them already synchronized.

**Comment by patrickdalla** [comment]
A similar behaviour occurs.
When the metadata is read, the setTimezone is called, but not applied. When the second setTimezone to UTC is called to 'normalize' the metadata timezone information, the first wasn't applied, so, the field are not shifted.

File: tika-parsers/tika-parsers-standard/tika-parsers-standard-modules/tika-parser-xmp-commons/src/test/java/org/apache/tika/parser/xmp/JempboxExtractorTest.java [hunk 1]
```diff
@@ -134,4 +134,14 @@ public void testMaxXMPMMHistory() throws Exception {
...
     }
 
+    @Test
+    public void testModifiedTZ() throws Exception {
+        Metadata m = new Metadata();
+        JempboxExtractor ex = new JempboxExtractor(m);
+        try (InputStream is = getResourceAsStream("/test-documents/testXMP.xmp")) {
+            ex.parse(is);
+        }
+        assertEquals("2014-03-04T22:50:41Z", m.get(XMPMM.HISTORY_WHEN));
```
[the remark this answers is not in the document]
Comment (patrickdalla) [reply tika-parsers/tika-parsers-standard/tika-parsers-standard-modules/tika-parser-xmp-commons/src/test/java/org/apache/tika/parser/xmp/JempboxExtractorTest.java:144]: Yes. It seems the desired. To normalize the string to UTC, making the correct time shifting. So, if it is working, that should really solve the issue.

[the remark this answers is not in the document]
Comment (patrickdalla) [reply tika-parsers/tika-parsers-standard/tika-parsers-standard-modules/tika-parser-xmp-commons/src/test/java/org/apache/tika/parser/xmp/JempboxExtractorTest.java:144]: How could I download a tika app snapshot with this patch to test on my PDF file?

**Comment by tballison** [comment]
> Yes, it seems you misunderstood (or I misinformed) the issue.
> 
> There is a XMP metadata tag in the PDF with the values: <rdf:Description rdf:about="" **xmp:ModifyDate="2023-09-06T13:35:38-04:00"** xmlns:xmp="[http://ns.adobe.com/xap/1.0/">](http://ns.adobe.com/xap/1.0/%22%3E)
> 
> Look that the timezone is -04:00. But tika does not extracts the timezone, parsing the value as "2023-09-06T13:35:38Z", as it was in UTC.

When I add the date string from your original issue to the testXMP.xmp file, I get this from Tika: 
```xmp:ModifyDate : 2023-09-06T17:35:38Z```

That's what you'd expect, right?

**Comment by patrickdalla** [comment]
yes. thanks

Em seg., 11 de set. de 2023 17:37, Tim Allison ***@***.***>
escreveu:

**Pull request merged**